config/common.go

```diff
@@ -34,6 +37,14 @@ type Unmarshallable interface {
 	Unmarshal(component *confmap.Conf) error
 }
 
+// Marshallable defines an optional interface for custom configuration marshalling.
+// A configuration struct can implement this interface to override the default marshalling.
+type Marshallable interface {
+	// Marshal is a function that marshals the config into an interface to be encoded by
+	// the YAML encoder. Provides the configencoder.Encoder.
+	Marshal(encoder configencoder.Encoder) (*confmap.Conf, error)
```

Comment:
Why not passing a `*confmap.Conf` and the idea is to "Marshal into the Conf"? Why is an encoder needed?

Reply:
The encoder is there to assist the user with marshaling their config. It gives the config a starting point to marshal the config and is optional to use.
```
func (t TestMarshallableExtension) Marshal(enc configencoder.Encoder) (*confmap.Conf, error) {
	out, err := enc.Encode(t)
	if err != nil {
		return nil, err
	}
	data, ok := out.(map[string]interface{})
	if !ok {
		return nil, errors.New("unable to encode struct")
	}
	conf := confmap.NewFromStringMap(data)
	conf.Delete("omit")
	return conf, nil
}
```
Right now, the marshaler doesn't encode the string map it receives from `confmap.Conf` after calling this function. Maybe it makes more sense to allow the config to just marshal into the `confmap.Conf` and then encode that result in the marshaler after.

The `confmap.Conf` doesn't have many functions enabling a user to change the underlying map, so I imagine the user would need to make the changes before creating the `confmap.Conf` with `NewFromStringMap`.

Reply:
@bogdandrutu I'm not opposed to passing in a `*confmap.Conf` instead, but how could a user marshal into it? It doesn't look like the underlying `koanf` instance provides any functions for modification aside from `Merge`/`Delete`.